src/models.rs

```diff
@@ -26,5 +28,68 @@ impl From<User> for UserWithoutSecrets {
 pub struct UserWithoutSecrets {
     pub username: String,
     pub balance: f64,
+    pub level: i32,
     pub collected_timestamp: SystemTime,
 }
+
+pub struct UpgradeableLevelsOutput {
+    pub level: i32,
+    pub cost: f64,
+}
+
+pub trait GameCalculations {
+    const BASE_COST: f64;
+    const BASE_PRODUCTION: f64;
+    const COST_MULTIPLIER: f64;
+    const PRODUCTION_MULTIPLIER: f64;
+
+    fn get_production(&self) -> f64;
+    fn next_level_cost(&self) -> f64;
+    fn specific_level_cost(&self, level: i32) -> f64;
+    fn upgradeable_levels(&self) -> Result<UpgradeableLevelsOutput, ()>;
+}
+
+impl GameCalculations for User {
+    const BASE_COST: f64 = 5.0;
+    const BASE_PRODUCTION: f64 = 1.0;
+    const COST_MULTIPLIER: f64 = 1.15;
+    const PRODUCTION_MULTIPLIER: f64 = 1.1;
+
+    /// calculates the production of the user
+    fn get_production(&self) -> f64 {
+        let mut production = Self::BASE_PRODUCTION;
+        for progressed_level in 1..self.level {
```

Comment:
is this supposed to be a `..=` or is it supposed to cut off right before the user's actual level?

Reply:
It's supposed to be `..=`, my bad